src/background/deinflect.ts

```diff
@@ -918,30 +87,34 @@ const enum Type {
   // Intermediate types
   Initial = 1 << 6, // original word before any deinflection (from-type only)
   VNai = 1 << 7,
-  Te = 1 << 8,
+  TaTeStem = 1 << 8,
+  DaDeStem = 1 << 9,
+  MasuStem = 1 << 10,
 }
 
 export { Type as WordType };
 
-interface DeinflectRule {
+type DeinflectRule = {
   from: string;
   to: string;
   // A bit mask representing the type of words to which this rule can be
   // applied.
   //
-  // For example, 遊びすぎる would match the びすぎる→ぶ rule where the from
-  // type is an ichidan/ru-verb while the to type is a godan/u-verb.
+  // For example, 遊びすぎる would match the "すぎる"→"" rule where the from type
+  // is an ichidan/ru-verb while the to type is the intermediate type masu-stem.
+  // The remaining 遊び would then match the "び"→"ぶ" rule where the from type
+  // is the intermediate type masu-stem while the to type is a godan/u-verb.
+  //
+  // Intermediate types ensure that some rules are only applied in specific
+  // conditions.
   //
-  // The fromType may also have the special value Initial which means it accepts
-  // anything BUT one of the other word types.
+  // For example, consider the deinflection rule "ます"→"" without the to type
+  // masu-stem. Then everything following, not just the correct masu stem, would
+  // match it. "食べろます" would be parsed as "< imperative < polite", which
+  // is obviously not correct.
   //
-  // For example, consider the deinflection rule that allows 食べ (imperative)
-  // to be de-inflected to 食べる: べ→べる. In this case, the to-type is an
-  // ichidan/ru-verb, while the from type is basically anything but NOT the
-  // result of any other deinflection (since they never produce verb stems).
-  // For this case we set the from-type to Initial so that it does NOT match
-  // any of the existing word types but it DOES match when we compare with 0xff
-  // (the type we use for the initial input).
+  // Similarly, the type Intial ensures that a rule is only applied when it is
+  // the last part of an inflecting word, as every to type is different from it.
```

Comment:
Is is the last part? I thought it was the first part?

Reply:
Well, depending on the perspective, I guess. It's the first part getting parsed but the last part structurally, i.e. the end of the word.

Reply:
I will clarify what is meant.